### PR TITLE
Switch PlanDevices pass to be w.r.t. SEScopes instead of DLDeviceTypes.

### PR DESCRIPTION
**Pull request by mbs-octoml**
CAUTION: Breaking VM executable serialization change. I needed a new 'virtual devices' array in the executable so that instructions can continue to refer to devices by a simple index yet the VM can respect both the device type and id for runtime devices. 

Continuing from #9313, and as part of apache/tvm-rfcs#38, we switch PlanDevices to plan with respect to SEScopes instead of just DLDeviceTypes. Our ultimate goal is to be able to flow memory scopes between PrimFuncs by re-running PlanDevices after the LowerTE pass. This PR at least gets us to being able to flow the memory scopes, but the actual changes to PlanDevices to look inside PrimFuncs is still two PR's in the future.

However, we get two nice side effects right away:
  - Since SEScopes contain Targets we can isolate all the device-to-target resolution machinery within PlanDevices (with the help of CompilationConfig). After PlanDevices has run we can retrieve the Target for any sub-expression directly from that sub-expression's SEScope. For now we retain the one-Target-per-DLDeviceType constraint since it baked into the public 'TargetMap' API, but the path to breaking that constraint is clearer.
  - Device ids are now respected all the way from annotation to executor. Previously though we had a bit of plumbing using Devices the device_id therein was ignored or defaulted to zero.  The Python "on_device" annotation helpers still work w.r.t. devices. Thus though they now respect device ids, they do not allow the user to specify a Target or memory scope as supported by the underlying SEScope.

### CODE REVIEW

**Comment by mbs-octoml** [comment]
(Nearly ready for review -- using ci to tease out what i've broken)

File: src/relay/backend/build_module.cc [hunk 1]
```diff
@@ -207,7 +210,7 @@ class RelayBuildModule : public runtime::ModuleNode {
     } else if (name == "optimize") {
       return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
         ICHECK_EQ(args.num_args, 2);
-        *rv = this->Optimize(args[0], args[1], this->params_);
```
Comment (jroesch) [inline src/relay/backend/build_module.cc:210]: Why do we no longer need the parameters here?

File: src/relay/backend/te_compiler.h [hunk 2]
```diff
@@ -173,31 +173,28 @@ Map<Target, IRModule> GetPerTargetModules(IRModule mod);
...
  * \param memory_plan The memory plan used during lowering
  * \param module_name The name of this module
  * \param process_fn Callback allowing one-level up code generators to process
  * each function that we lower
  * \return The lowered module, see above.
  */
 IRModule LowerTE(
-    const IRModule& module, TargetMap targets, backend::StaticMemoryPlan memory_plan,
-    const String& module_name, ProcessFn process_fn = [](Function f) {});
+    const IRModule& module, backend::StaticMemoryPlan memory_plan, const String& module_name,
```
Comment (jroesch) [inline src/relay/backend/te_compiler.h:183]: Do we have a tracking issue for pulling out the mangling from the lowering? probably makes sense after your most recent draft change?

File: tests/python/unittest/test_micro_model_library_format.py [hunk 3]
```diff
@@ -411,4 +410,7 @@ def test_export_byoc_c_module():
 
 
 if __name__ == "__main__":
-    sys.exit(pytest.main([__file__] + sys.argv[1:]))
+    import sys
+
+    # sys.exit(pytest.main([__file__] + sys.argv[1:]))
+    test_export_operator_model_library_format()
```
Comment (jroesch) [inline tests/python/unittest/test_micro_model_library_format.py:416]: We should probably fix this, feel free to do in follow up.

**Review by jroesch — APPROVED** [review]
Okay mostly looks mechanical to me, LGTM.

File: src/relay/backend/build_module.cc (hunk 1 above)
Comment (mbs-octoml) [reply src/relay/backend/build_module.cc:210]: The code was using both local and member params simultaneousy, but the local was only ever bound to the member anyway. So I just got rid of the local.

Note there's some weirdness in this API as to what state is expected to come over between 'build' and 'optimize'.

File: src/relay/backend/te_compiler.h (hunk 2 above)
Comment (mbs-octoml) [reply src/relay/backend/te_compiler.h:183]: That's CORE-47 I believe.

File: tests/python/unittest/test_micro_model_library_format.py (hunk 3 above)
Comment (mbs-octoml) [reply tests/python/unittest/test_micro_model_library_format.py:416]: Woops, thanks.

**Comment by mbs-octoml** [comment]
I'm trying to repro the tests/python/driver/tvmc/test_compiler.py::test_compile_tflite_module_with_external_codegen_cmsisnn failure.

Note https://github.com/apache/tvm/pull/9480, which changed the assert for 3 to 4 artifacts, while I'm still getting 3. That's suspicions.

Attempting to repro in the docker ci-cpu image terminates my remote desktop session (?!?). I'll try to get the cnsis-nn setup going directly, which is a good chance to experience the BYOC dependency setup first hand.

**Comment by mbs-octoml** [comment]
Ready to merge.

**Pull request merged**